### ISSUE TITLE
Improve PrimitiveColumn.Round() to take in the number of decimal places

### ISSUE THREAD

**Issue by pgovind** [dotnet/corefxlab#2661]

**Comment by RamonWill** [comment]
Hi @pgovind ,

Could i help on this issue?

If so could a potential solution be to add in an optional argument to Round such as "decimalPlace" that will determine the decimal places to round by

```diff
        public void Round(PrimitiveColumnContainer<double> column, int decimalPlaces=0)
        {
            for (int b = 0; b < column.Buffers.Count; b++)
            {
                var buffer = column.Buffers[b];
                var mutableBuffer = DataFrameBuffer<double>.GetMutableBuffer(buffer);
                var mutableSpan = mutableBuffer.Span;
                for (int i = 0; i < mutableSpan.Length; i++)
                {
                    mutableSpan[i] = (double)(Math.Round((decimal)mutableSpan[i], decimalPlaces));
                }
                column.Buffers[b] = mutableBuffer;
            }
        }
```

Kind Regards,

Ramon

**Comment by pgovind** [comment]
> Could i help on this issue?

@RamonWill: Yes of course! You are on the right track with your idea, but I think the better way is to modify the existing `Round` method [here](https://github.com/dotnet/corefxlab/blob/4e823aa0178c9aeff2bcf7de65ce27efc743ffbb/src/Microsoft.Data.Analysis/DataFrameColumn.Computations.cs#L170). You can't change that file directly though, because it is a text-template generated file. Unfortunately, this area of the code is not very approachable. So you'd have to modify 
1) The [DataFrameColumn.Computations.tt](https://github.com/dotnet/corefxlab/blob/master/src/Microsoft.Data.Analysis/DataFrameColumn.Computations.tt) file to produce `Round(bool inPlace = false, int decimalPlaces = 0)` first and then follow the implementation code to 
2) Change (PrimitiveDataFrameColumnComputations)[https://github.com/dotnet/corefxlab/blob/master/src/Microsoft.Data.Analysis/PrimitiveDataFrameColumnComputations.tt] similarly to produce `void Round(PrimitiveColumnContainer<T> column, int decimalPlaces = 0);` and modify the existing `Round` implementation in each `<ColumnType>Computation` class.

I can help you through each stage if you'd like. Or if you'd rather tackle a more focused issue, I can find some for you :)

**Comment by RamonWill** [comment]
Hi @pgovind if you could help through each stage that would be really appreciated whenever you have the time :). I think this project is a great idea and i would love to help out when I can.

Looking through the steps you posted:

> modify The DataFrameColumn.Computations.tt file to produce Round(bool inPlace = false, int decimalPlaces = 0) first

From what i can gather in the [ColumnsArtihmetricTemplate.ttinclude](https://github.com/dotnet/corefxlab/blob/master/src/Microsoft.Data.Analysis/ColumnArithmeticTemplate.ttinclude#L206-L228) file an array of MethodConfiguration objects are created.
Then in the [DataFrameColumnComputations.tt](https://github.com/dotnet/corefxlab/blob/master/src/Microsoft.Data.Analysis/DataFrameColumn.Computations.tt#L21) file the MethodConfiguration objects are then made into methods for the DataFrameColumn.Computations.cs file based on the if/else if/else statements.

> then follow the implementation code  Change PrimitiveDataFrameColumnComputations similarly to produce void Round(PrimitiveColumnContainer<T> column, int decimalPlaces = 0); and modify the existing Round implementation in each <ColumnType>Computation class.

the steps here are similar to the above DataFrameColumn setup.

Is my thinking correct? If it is, what would need to be changed in the DataFrameColumnComputations.tt file because i can see that currently Abs, CumulativeMax, CumulativeMin, CumulativeProduct, CumulativeSum all have the same condition as Round().

Thank you for your time,
Ramon

**Comment by RamonWill** [comment]
Hi @pgovind ,
Was just following up on the above. Would you be able to provide guidance?

I'd also be very happy to help on more focused issues you have for this project :)
Kind regards,
Ramon

**Comment by pgovind** [comment]
Apologies, I must've missed your update. Your understanding is correct. Here's what I would do:

1. Add a new public field called `ExtraArguments` to the MethodConfiguration class
2. Add a default `string extraArguments = ""` parameter to the MethodConfiguration constructor and initialize the field.
3. Update the `new MethodCOnfiguration(Round, ...` line in ColumnArithmeticTemplate.ttinclude by passing in the argument `extraArguments = ", int decimalPlaces = 0`.
4. Update both DataFrameColumnComputations.tt and  PrimitiveDataFrameColumnComputations to something like:
```
<# } else if (compMethod.MethodType == MethodType.ElementwiseComputation && compMethod.HasReturnValue == false) {#>
        public virtual DataFrameColumn <#=compMethod.MethodName#>(bool inPlace = false<#=compMethod.ExtraArguments#>) // Note that "," is inside extraArguments
```

That should produce the correct `Round(bool inPlace = false, int decimalPlaces = 0)` signature

**Comment by RamonWill** [comment]
much appreciated for this @pgovind that all makes sense. I will make a PR for this soon as well as a test.

Kind Regards,

Ramon

**Comment by RamonWill** [comment]
Hi @pgovind 

I came across a slight issue arising from [PrimitiveDataFrameColumn.Computations.tt](https://github.com/dotnet/corefxlab/blob/1d9a590490664933add4a43d698f78af298aaea2/src/Microsoft.Data.Analysis/PrimitiveDataFrameColumn.Computations.tt#L51). Somehow, I need to use the Extra Arguments in the below line. However, since the string contains a datatype and is hardcoded to equal zero it will cause an error.

```diff
<# } else if (compMethod.MethodType == MethodType.ElementwiseComputation && compMethod.HasReturnValue == false) {#>
            PrimitiveDataFrameColumn<T> ret = <#=GenerateInPlaceStatement("this", "Clone()")#>;
+            PrimitiveColumnComputation<T>.Instance.<#=compMethod.MethodName#>(ret._columnContainer<#=compMethod.ExtraArguments#>); // i have to add the extra arguments here but since the arguments contain the data type it causes an error.
            return ret;
```


I will end up with something like the following in the [corresponding cs file](https://github.com/dotnet/corefxlab/blob/master/src/Microsoft.Data.Analysis/PrimitiveDataFrameColumn.Computations.cs#L145)

```diff
        /// <inheritdoc/>
+        public override DataFrameColumn Round(bool inPlace = false, int decimalPlaces=0)
        {
            PrimitiveDataFrameColumn<T> ret = inPlace ? this : Clone();
+            PrimitiveColumnComputation<T>.Instance.Round(ret._columnContainer, int decimalPlaces=0); // issue here
            return ret;
        }
```
I do have a thought of maybe making a Class for Extra Arguments and then it has a method that will create a string with the datatypes and default values for each argument and another method that would return a string with all the arguments without the default values and datatypes.

i.e. one method would return "int decimalPlaces=0, ..., ..." and the other would just return "decimalPlaces, ..., ..."

Do let me know your thoughts i could be overcomplicating the issue.

Kind Regards,

Ramon
